streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java

```diff
@@ -216,5 +217,47 @@ private V fetchPrevious(final Bytes key, final long timestamp) {
             }
         }
     }
+    
+    @Override
+    public KeyValueIterator<Windowed<Bytes>, byte[]> all() {
+        validateStoreOpen();
 
+        final KeyValueIterator<Windowed<Bytes>, byte[]>  underlyingIterator = underlying.all();
+        final ThreadCache.MemoryLRUCacheBytesIterator cacheIterator = cache.all(name);
+
+        return new MergedSortedCacheWindowStoreKeyValueIterator(
+            cacheIterator,
+            underlyingIterator,
+            bytesSerdes,
+            windowSize,
+            cacheFunction
+        );
+    }
+    
+    @Override
+    public KeyValueIterator<Windowed<Bytes>, byte[]> fetchAll(long timeFrom, long timeTo) {
+        validateStoreOpen();
+        
+        final KeyValueIterator<Windowed<Bytes>, byte[]> underlyingIterator = underlying.fetchAll(timeFrom, timeTo);
+        final ThreadCache.MemoryLRUCacheBytesIterator cacheIterator = cache.all(name);
+        
+        final NavigableSet<Bytes> allBytes = cache.keySet(name);
```

Comment:
Oh, this was never changed because it is not neccesarily the first key and last key of a ```NamedCache``` which encompasses the allocated time slot. By that, I mean that this is in the ```fetchAll``` method, not the ```all``` method. Hence, we could not just simply adopt the first and last key for an acceptable key range. And that is why those two methods you mentioned above was never used.

Reply:
I originally did not mean to change the logic, but just the code hierarchy: i.e. not exposing a `keySet` function of `ThreadCache`.

But looking at this logic a second time, I'm not sure if it's logic is correct either: note that for the window store's cached keys, it is in the form of `[segmentId, key, timestamp, seqNum]`, and the navigable set of the bytes is sorting the keys in lexicographical order, so the `allBytes.floor(byte[0]. timeTo, 0)` would not take the segment id into consideration, right?

Reply:
Yeah, I didn't think about that. 
I have come up with another approach using ```HasNextCondition``` since before we did not have the ```null``` feature in ```KeySchema```'s ```hasNextCondition()``` method.